### DISCUSSION TITLE
swich tasmota from thingsboard

### DISCUSSION THREAD

**Discussion by RepiZoli** [arendst/Tasmota#15565] [Q&A] [answered]
Hi I new in Tasmota, my aim is to connect tasmota with thingsboard. In case of telemetry (where I send data from tasmota to thingsboard (publish to thingsboard))it worked. However in case  I want to switch tasmota from thingsboard is not working. 
I can subscribe to the following: v1/devices/me/rpc/request/+ with :
 19:55:29.669 CMD: Subscribe2 kapcs, v1/devices/me/rpc/request/+ 
19:55:29.678 MQT: esp/stat/RESULT = {"Subscribe":"KAPCS,v1/devices/me/rpc/request/+"}
the time out failure disappear on the thingsbord dashboard.
 But as I try to swich I get a failure for each and every switch

19:55:33.661 MQT: esp/stat/RESULT = {"Command":"Unknown"}
19:55:34.161 MQT: esp/stat/RESULT = {"Command":"Unknown"}

According to the Thingsboard:

Server-side RPC
In order to subscribe to RPC commands from the server, send SUBSCRIBE message to the following topic:

v1/devices/me/rpc/request/+
Once subscribed, the client will receive individual commands as a PUBLISH message to the corresponding topic:

v1/devices/me/rpc/request/$request_id
where $request_id is an integer request identifier.

According to my understanding thingboard publish to a different topic.

I also found a python script for this: 

Here the it is subscribe to the : v1/devices/me/rpc/request/+ 

In the script this problem is bridged with the following method:
if msg.topic.startswith('v1/devices/me/rpc/request/'):

 And the system working with it:

Topic: v1/devices/me/rpc/request/98Message: b'{"method":"setValue","params":false}'

My question is:
1. How can I get this message in tasmota
2. How can I convert the params:"false" to Power= off / true  to Power= On

Thanks in advance ( the pyton scriot below)

# This Program illustrates the Simple Server Side RPC on ThingsBoard IoT Platform
# Paste your ThingsBoard IoT Platform IP and Device access token
# RPC Server_Side_RPC.py : This Program will illustrates the Server side RPC
import os
import time
import sys
import json
import random
import paho.mqtt.client as mqtt

# Thingsboard platform credentials
THINGSBOARD_HOST = 'demo.thingsboard.io'
ACCESS_TOKEN = ''  
button_state = {"enabled": False}

def setValue(params):
    button_state['enabled'] = params
    print("Rx setValue is : ", button_state)

# MQTT on_connect callback function
def on_connect(client, userdata, flags, rc):
    print("rc code:", rc)
    client.subscribe('v1/devices/me/rpc/request/+')

# MQTT on_message caallback function
def on_message(client, userdata, msg):
    print('Topic: ' + msg.topic + 'Message: ' + str(msg.payload))
    if msg.topic.startswith('v1/devices/me/rpc/request/'):
        requestId = msg.topic[len('v1/devices/me/rpc/request/'):len(msg.topic)]
        print("requestId : ", requestId)
        data = json.loads(msg.payload)
        if data['method'] == 'getValue':
            print("getvalue request\n")
            print("sent getValue : ", button_state)
            client.publish('v1/devices/me/rpc/response/' + requestId, json.dumps(button_state), 1)
            print(msg)
        if data['method'] == 'setValue':
            print('Zoli')
            print("setvalue request\n")
            params = data['params']
            setValue(params)
            client.publish('v1/devices/me/attributes', json.dumps(button_state), 1)

# start the client instance
client = mqtt.Client()

# registering the callbacks
client.on_connect = on_connect
client.on_message = on_message

client.username_pw_set(ACCESS_TOKEN)
client.connect(THINGSBOARD_HOST, 1883, 150)

try:
    client.loop_forever()

except KeyboardInterrupt:
    client.disconnect()

**Comment by barbudor** [comment, marked as answer]
Using [`Subscribe`](https://tasmota.github.io/docs/Commands/#subscribe)
`Subscribe power,v1/devices/me/rpc/request/98,params` will create an event named `power` which subscribe to the given topic and get the value of the key `params`
You should be able to get it with a [Rule](https://tasmota.github.io/docs/Rules)
```
Rule1 ON event#power=false DO Power OFF ENDON ON event#power=true DO Power ON ENDON
Rule1 1
```

**Reply by ztzintelligence** [reply]
@RepiZoli Hi, Can you share how to tasmota publish data to thingsboard, i am try to do this, but failed. Because the thingsboard  need a topic like v1/devices/me/telemetry, how to configure tasmota to match this need?

i modify the prefix and i got output like this:
![image](https://user-images.githubusercontent.com/82553410/167294261-2532999c-0b2c-4aa0-8925-78c0e99b5cfd.png)
and the topic just different from **v1/devices/me/telemetry**,
v1/devices/me/cmnd/POWER
v1/devices/me/telemetry/INFO1
v1/devices/me/telemetry/INFO2
v1/devices/me/stat/POWER

So, how to get a valid topic thingsboard need: **v1/devices/me/telemetry**

**Reply by barbudor** [reply]
I would try
```
Topic me
FullTopic v1/devices/%topic%/%prefix%/
Prefix3 telemetry
```

However if I understand the way Things Board is working, I'm not sure you will get full Tasmota support with Thingsbloard borked mqtt server.

**Reply by ztzintelligence** [reply]
@barbudor Thanks for reply, i modify the tasmota source code, and finally i can publish data to thingsboard. Your advise will try late.

**Comment by RepiZoli** [comment]
Hi all

For to publish I used the following rule:
Rule2
  ON tele-DHT11#temperature DO publish v1/devices/me/telemetry {"temperature":"%value%"} ENDON

@barbudor 

Thanks my problerm is that according to the thingsboard you have to subscribe v1/devices/me/rpc/request/+ and than you get a respond from the thingsboard on the folloing topic:

v1/devices/me/rpc/request/$request_id

$request_id  change continously so I can not subscribe to the following topic: v1/devices/me/rpc/request/98 

But I will try

**Reply by barbudor** [reply]
@RepiZoli
OK, I tougth the 98 was a class of response, like 98=power, 99=something elsd
In that case it won't work with rules as you can know what is the part of the topic matching the +. But if you can ignore that may it will work
Alternatively on ESP32 with Berry language you would have more flexibility to manage topics and payload.

**Reply by RepiZoli** [reply]
@barbudor 
 HI meanwhile I read that the + is a wildcard
http://www.steves-internet-guide.com/understanding-mqtt-topics/

Using Topic Wildcards
Subscribing to topic house/#

Covers

house/room1/main-light
house/room1/alarm
house/garage/main-light
house/main-door
house/
house
etc
Subscribing to topic house/+/main-light

covers

house/room1/main-light
house/room2/main-light
house/garage/main-light

'-----------------------

I also checked with MQTT BOX:

Subscribe to v1/devices/me/rpc/request/+
respond was the follow: 

{"method":"setValue","params":false}
qos : 1, retain : false, cmd : publish, dup : false, topic : v1/devices/me/rpc/request/157, messageId : 7, length : 69

Maybe this help? Subscribe2 power,v1/devices/me/rpc/request/+,params

**Reply by barbudor** [reply]
Yes + is a one-level wildcard and # is a multi-level wildcard. That's part of mqtt basics

By subscribing to v1/devices/me/rpc/request/+ a Tasmota event will be triggered whatever is the request ID. But the request itself will not be provided to the rule, only the payload.
So that should be fine if you don't care about the ID and only about the payload
You should be able to use either 'Subscribe' with v1/devices/me/rpc/request/+ or Subscribe2 with v1/devices/me/rpc/request as the later automatically adds the # wildcard at the end (don't include the + in that case) 

If you try something and feel it doesn't work, we will need more details
Exact subscribe command (result of subscribe without parameters should provide the current subscription) 
Exact rules (result of command rule0) 
Console log after Weblog 3 to see more details 
Exact rule
